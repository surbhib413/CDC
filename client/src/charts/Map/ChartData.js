import React, { useState, useEffect } from "react";
import styles from "./Map.module.scss";

const ChartData = (props) => {
  const [loading, setLoading] = useState(false);

  const { chartData } = props;
  //console.log(chartData.length);
  const renderProgressBar = (regionCode, term, spot, ind) => {
    //console.log(document.getElementById("progressBarSet"));
    // console.log(regionCode);
    if (
      document.querySelector(`#progressbarDataForTerm${regionCode}`) !== null
    ) {
      document.querySelector(
        `#progressbarDataForTerm${regionCode}`
      ).style.width = `${term / 10}%`;

      document.querySelector(
        `#progressbarDataForTerm${regionCode}`
      ).innerHTML = `<span id="hsTerm${regionCode}" style="font-size:16px;float:left; font-style: italic;">${term}</span><span id="lsTerm${regionCode}" style="font-size:16px;float:right; font-style: italic;left:150px;position:absolute">${
        1000 - term
      }</span>`;
    }
    if (
      document.querySelector(`#progressbarDataForSpot${regionCode}`) !== null
    ) {
      document.querySelector(
        `#progressbarDataForSpot${regionCode}`
      ).style.width = `${spot / 10}%`;
      document.querySelector(
        `#progressbarDataForSpot${regionCode}`
      ).innerHTML = `<span id="hsSpot${regionCode}" style="font-size:16px;float:left; font-style: italic;">${spot}</span><span id="lsSpot${regionCode}" style="font-size:16px;float:right; font-style: italic;left:150px;position:absolute">${
        1000 - spot
      }</span>`;
    }

    if (
      document.querySelector(`#progressbarDataForIndigenous${regionCode}`) !==
      null
    ) {
      document.querySelector(
        `#progressbarDataForIndigenous${regionCode}`
      ).style.width = `${ind / 10}%`;
      document.querySelector(
        `#progressbarDataForIndigenous${regionCode}`
      ).innerHTML = `<span id="hsIndigenous${regionCode}" style="font-size:16px;float:left; font-style: italic;">${ind}</span><span id="lsIndigenous${regionCode}" style="font-size:16px;float:right; font-style: italic;left:150px;position:absolute">${
        1000 - ind
      }</span>`;
    }
  };

  useEffect(() => {
    setLoading(true);
  }, [props.chartData]) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <div>
      <div className={`${styles.refineryLocale}`}>
        <span>{props.location.toUpperCase()}</span>
        <div className={`${styles.progressbarTerm}`}>
          <div className={`${styles.progressbarData}`}></div>{" "}
        </div>
      </div>
      {loading && chartData.length > 0 ? (
        chartData.map((data, index) => {
          // console.log(data);
          let regionStyle = data.region.replace(/\s/g, "").toLowerCase();
          // console.log(regionStyle);
          setTimeout(function () {
            renderProgressBar(
              data.regionCode,
              data.term_hs,
              data.spot_hs,
              data.indigenous_hs
            );
          }, 300);
          if (styles[regionStyle]) {
            return (
              <div
                key={index}
                className={`${styles[regionStyle]}`}
                id='progressBarSet'
              >
                <div className={`${styles.chartDataFont}`}>
                  {data ? data.overall_percentage + "%" : 0}
                </div>
                <div style={{ fontSize: "24px" }}>
                  {data ? data.region : ""}{" "}
                </div>

                <div
                  className={`${styles.progressbarTerm}`}
                  id={`termText${data.regionCode}`}
                >
                  <div
                    className={`${styles.progressbarData}`}
                    id={`progressbarDataForTerm${data.regionCode}`}
                  ></div>{" "}
                </div>

                <div
                  className={`${styles.progressbarSpot}`}
                  id={`spotText${data.regionCode}`}
                >
                  <div
                    className={`${styles.progressbarData}`}
                    id={`progressbarDataForSpot${data.regionCode}`}
                  ></div>
                </div>
                <div
                  className={`${styles.progressbarIndigenous}`}
                  id={`indText${data.regionCode}`}
                >
                  <div
                    className={`${styles.progressbarData}`}
                    id={`progressbarDataForIndigenous${data.regionCode}`}
                  ></div>
                </div>
              </div>
            );
          }
        })
      ) : (
        <div>Loading....</div>
      )}
    </div>
  );
};

export default ChartData;
