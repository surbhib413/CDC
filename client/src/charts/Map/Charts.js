import React, { useState, useEffect } from "react";
import Chart from "react-google-charts";
import BackButton from "./BackButton";
import Table from "./Table";
import CountryChart from "./CountryChart";
//import _ from "lodash";
import styles from "./Map.module.scss";
import ChartData from "./ChartData";
import SelectBox from "./SelectBox";
import LegendsProgressBar from "./LegendsProgressBar";
import env from "../../environment";

const GoogleChart = (props) => {
  const [data, setData] = useState([]);
  const [chartData, setChartData] = useState([]);
  const [loading, setLoading] = useState(false);
  const [selectBoxMode, setSelectBoxMode] = useState("ytm");
  const [location, setLocation] = useState(props.location);

  const [childData, setChildData] = useState({
    clicked: false,
    regionClicked: {},
  });

  const mapClickHandler = (e, region) => {
    if (data[region + 1]) {
      let fetchedRegionName = data[region + 1][0];
      const bool = chartData.map((item) => {
        item.countries.map((el) => {
          if (el.v === fetchedRegionName) {
            //console.log(item);
            props.getRegionNameOnClick(item.region.toLowerCase());
            setChildData({
              clicked: !clicked,
              regionClicked: item,
            });
          }
        });
      });
    }
  };
  let temp = [];
  const storeCountry = (
    countryList,
    percent,
    region,
    regionCode,
    term_hs,
    spot_hs,
    indigenous_hs
  ) => {
    countryList.map((country) => {
      temp.push([country.v, percent]);
      setData([["Country", "Percent"], ...temp]);
    });

    setChartData((state) => [
      ...state,
      {
        region: region,
        regionCode: regionCode,
        countries: countryList,
        overall_percentage: percent,
        term_hs,
        spot_hs,
        indigenous_hs,
      },
    ]);
  };

  const fetchData = () => {
    let url = `${env.ITRM_KPI_URL}/crude_procurement/region?refinery=${props.location}&data_type=${selectBoxMode}`;
    fetch(url, { method: "GET" })
      .then((res) => res.json())
      .then((response) =>
        response.data.map((el, index) =>
          storeCountry(
            el.country_list,
            el.overall_percentage,
            el.region,
            el.regionCode,
            el.term_hs,
            el.spot_hs,
            el.indigenous_hs
          )
        )
      )
      .catch((error) => console.log(error));
  };
  useEffect(() => {
    fetchData();
    setLoading(true);
    setLocation(props.location);
    setChartData([]);
  }, [selectBoxMode, props.location]) // eslint-disable-line react-hooks/exhaustive-deps

  const backClickHandler = (e, region) => {
    setChildData({ clicked: !clicked });
  };
  const chartEvents = [
    {
      eventName: "ready",
      callback: ({ chartWrapper, google }) => {
        const chart = chartWrapper.getChart();
        chart.container.addEventListener("click", (ev) => {
          if (!clicked) {
            const indexOfSelectedRegion =
              chart.getSelection().length > 0
                ? chart.getSelection()[0]["row"]
                : null;
            if (indexOfSelectedRegion) {
              mapClickHandler(ev, indexOfSelectedRegion);
            } else {
              return false;
            }
          }
        });
      },
    },
  ];
  const { regionName, dataWeight, clicked } = childData;
  const { regionClicked } = childData;

  const getSelectBoxMode = (val) => {
    setSelectBoxMode(val);

    // setTimeout(function() {
    //   fetchData();
    // }, 2000);
  };
  //console.log(data);
  return (
    <div>
      {clicked ? (
        <div>
          <BackButton backClickHandler={() => backClickHandler()} />
          <CountryChart regionClicked={regionClicked}></CountryChart>
        </div>
      ) : loading ? (
        <div style={{ position: "relative" }}>
          <div className={`${styles.crudeProcurementHeading}`}>
            Crude Procurement(MMT)
          </div>
          <Chart
            width={"100%"}
            chartType='GeoChart'
            chartEvents={chartEvents}
            options={{
              colorAxis: {
                colors: [
                  "#FF7800",
                  "#ffbf00",
                  "#ffec00",
                  "#fff25d",
                  "#fff690",
                  "#fffabd",
                ],
              },
              backgroundColor: "#000000",
              datalessRegionColor: "#283040",
              defaultColor: "#000000",
              stroke: "#6699cc",
              legend: "none",
              tooltip: { trigger: "none" },
            }}
            data={data}
            mapsApiKey={process.env.REACT_APP_GOOGLE_KEY}
            style={{
              transform: `scale(${0.8})`,
              position: "absolute",
              right: "100px",
              cursor: "pointer",
            }}
          />
          <div className={`${styles.selectbox}`}>
            <SelectBox getSelectBoxMode={(val) => getSelectBoxMode(val)} />
          </div>
          <LegendsProgressBar />
          <ChartData chartData={chartData} location={location} />

          <div className={`${styles.mapTable}`}>
            <div className={`${styles.table}`}>
              <Table regionName='all regions' />
            </div>
          </div>
        </div>
      ) : (
        <div>Loading...</div>
      )}
    </div>
  );
};
export default GoogleChart;
