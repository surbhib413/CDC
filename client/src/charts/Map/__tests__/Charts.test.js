import React from "react";
import { shallow, mount } from "enzyme";
import Charts from "../Charts";
import ChartData from "../ChartData";
import sinon from "sinon";

describe("Charts Component", () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    wrapper = mount(<Charts location='middle east' />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("charts component is enclosed withn GoogleChart component", () => {
    expect(wrapper.find("GoogleChart").length).toEqual(1);
  });
  it("should match the snapshot", () => {
    const wrapper = shallow(<Charts />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("consists of a child component Legends ProgressBar", () => {
    expect(wrapper.find("LegendsProgressBar").length).toEqual(1);
  });
});
