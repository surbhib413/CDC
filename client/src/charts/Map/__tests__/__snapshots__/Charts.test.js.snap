// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts Component should match the snapshot 1`] = `"<div><div>Loading...</div></div>"`;
